engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java

```diff
@@ -0,0 +1,508 @@
+              selfField,
+              otherField
+          );
+        }
+        if (!areFieldsSame) {
+          return false;
+        }
+      }
+    } else {
+      // If there are more fields than equalsNodeCountForFields, just bailout and use
+      // uncached variant of EqualsNode and InvokeEqualsNode
+      CompilerDirectives.transferToInterpreterAndInvalidate();
+      for (int i = 0; i < fieldsSize; i++) {
+        Object selfField = self.getFields()[i];
+        Object otherField = otherAtom.getFields()[i];
+        boolean areFieldsSame;
+        if (selfField instanceof Atom selfFieldAtom
+            && otherField instanceof Atom otherFieldAtom
+            && atomOverridesEquals(selfFieldAtom)) {
+          areFieldsSame = InvokeEqualsNode.getUncached().execute(selfFieldAtom, otherFieldAtom);
+        } else {
+          areFieldsSame = EqualsNodeGen.getUncached().execute(selfField, otherField);
+        }
+        if (!areFieldsSame) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Helper node for invoking `==` method on atoms, that override this method.
+   */
+  @GenerateUncached
+  static abstract class InvokeEqualsNode extends Node {
+    static InvokeEqualsNode getUncached() {
+      return InvokeEqualsNodeGen.getUncached();
+    }
+
+    static InvokeEqualsNode build() {
+      return InvokeEqualsNodeGen.create();
+    }
+
+    abstract boolean execute(Atom selfAtom, Atom otherAtom);
+
+    @Specialization(guards = "cachedSelfAtomCtor == selfAtom.getConstructor()")
+    boolean invokeEqualsCachedAtomCtor(Atom selfAtom, Atom otherAtom,
+        @Cached("selfAtom.getConstructor()") AtomConstructor cachedSelfAtomCtor,
+        @Cached("getEqualsMethod(cachedSelfAtomCtor)") Function equalsMethod,
+        @Cached ExecuteCallNode executeCallNode,
+        @CachedLibrary(limit = "3") InteropLibrary interop) {
+      assert atomOverridesEquals(selfAtom);
+      Object ret = executeCallNode.executeCall(
+          equalsMethod,
+          null,
+          State.create(EnsoContext.get(this)),
+          new Object[]{selfAtom, otherAtom}
+      );
+      try {
+        return interop.asBoolean(ret);
+      } catch (UnsupportedMessageException e) {
+        throw new IllegalStateException(e);
+      }
+    }
+
+    @Specialization(replaces = "invokeEqualsCachedAtomCtor")
+    boolean invokeEqualsUncached(Atom selfAtom, Atom otherAtom,
+        @Cached ExecuteCallNode executeCallNode) {
+      Function equalsMethod = getEqualsMethod(selfAtom.getConstructor());
+      Object ret = executeCallNode.executeCall(
+          equalsMethod,
+          null,
+          State.create(EnsoContext.get(this)),
+          new Object[]{selfAtom, otherAtom}
+      );
+      assert InteropLibrary.getUncached().isBoolean(ret);
+      try {
+        return InteropLibrary.getUncached().asBoolean(ret);
+      } catch (UnsupportedMessageException e) {
+        throw new IllegalStateException(e);
```

Comment:
Hide behind `transferToInterpreter` or `@TruffleBoundary`, it cannot be no fast path anyway as it calls `printStackTrace()`.